src/nvim/buffer.c

```diff
@@ -5475,6 +5480,17 @@ int buf_signcols(buf_T *buf)
         linesum++;
       }
     }
+
+    if (curline > 0) {
+      linesum += decor_signcols(buf, &decor_state, (int)curline-1, (int)curline-1);
+    }
+    if (linesum > signcols) {
+      signcols = linesum;
+    }
+
+    // Check extmarks between signs
+    linesum = decor_signcols(buf, &decor_state, 0, (int)buf->b_ml.ml_line_count-1);
```

Comment:
There is a problem here. Given that `buf_signcols` is called often, this means the entire tree has to be rescanned each time, sometimes on each line change (as `mark_adjust` will invalidate the signcol count). This is what the inherited code from vim does, but our code aims to scale much better to huge amount of marks.

I'd suggest, for now only keep track of whether the buffer does contain extsigns or not. This can be done with a counter similar to `buf->b_virt_line_blocks` , and then we say extsigns always contribute 1 to `signcols` when there is at least one the buffer. This will be good enough in many cases, including the default configuration. Then we can consider how to support `auto:[MANY]` effectively as a follow-up.

Reply:
Hmm, it's important to me that `auto` works properly. I'll go ahead with your suggestion, but I'd want to follow this up pretty quickly, preferably before release. Do you have any ideas on how we can do this more efficiently? My only idea would be to store sign extmarks in a different tree to avoid iterating over the buftext marks each time, they don't really benefit being stored in the same tree anyway.

Personally, I would be happy to take the perf hit. Is it really that bad?

Reply:
> Is it really that bad?

Yes, it defeats the purpose of having a B-tree, at least on the read side, if the entire tree is being read every time time there is a change.

> Personally, I would be happy to take the perf hit.

A compromise then is to as I suggested for yes, no, and `auto[:1]` , and the full scan only on `auto:bigger`. 

> My only idea would be to store sign extmarks in a different tree to avoid iterating over the buftext marks each time.

This makes sense. In general having multiple trees is something I'd like to support, also for virt_lines/anticonceal.